File: Objective-C Sample Code/Chapter 7/Going Further/Top10/Top10 WatchKit App Extension/ComplicationController.m
//
//  ComplicationController.m
//  Top10 WatchKit App Extension
//
//  Created by Cory Bohon on 12/5/16.
//  Copyright © 2016 Cory Bohon. All rights reserved.
//

#import "ComplicationController.h"

@interface ComplicationController ()

@property (nonatomic, strong) NSArray *timelineEntries;

@end

@implementation ComplicationController

- (id)init
{
    if (self = [super init]) {
        self.timelineEntries = @[@"Never Gonna Give You Up by Rick Astly", @"Unlock My iPhone by FBI", @"The Screen Door Song by Kyle Richter"];
    }
    return self;
}

- (void)requestedUpdateDidBegin
{
    
}

#pragma mark - Timeline Configuration 
- (void)getSupportedTimeTravelDirectionsForComplication:(CLKComplication *)complication withHandler:(void (^)(CLKComplicationTimeTravelDirections))handler
{
    handler(CLKComplicationTimeTravelDirectionForward);
}

- (void)getTimelineStartDateForComplication:(CLKComplication *)complication withHandler:(void (^)(NSDate * _Nullable))handler
{
    handler([NSDate date]);
}

- (void)getTimelineEndDateForComplication:(CLKComplication *)complication withHandler:(void (^)(NSDate * _Nullable))handler
{
    handler([[NSDate date] dateByAddingTimeInterval:6 * 60 * 60]);
}

- (void)getPrivacyBehaviorForComplication:(CLKComplication *)complication withHandler:(void (^)(CLKComplicationPrivacyBehavior))handler
{
    handler(CLKComplicationPrivacyBehaviorShowOnLockScreen);
}

#pragma mark - Timeline Population
- (void)getCurrentTimelineEntryForComplication:(CLKComplication *)complication withHandler:(void (^)(CLKComplicationTimelineEntry * _Nullable))handler
{
    if (complication.family == CLKComplicationFamilyModularLarge) {
        CLKComplicationTimelineEntry *entry = [self createModularLargeTimelineEntryWithHeaderText:@"Top 10" bodyText:[self.timelineEntries firstObject] andDate:[NSDate date]];
        handler(entry);
    }
    
    if (complication.family == CLKComplicationFamilyUtilitarianLarge) {
        CLKComplicationTimelineEntry *entry = [self createUtilitarianLargeTimelineEntryWithHeaderText:@"Top 10" bodyText:[self.timelineEntries firstObject] andDate:[NSDate date]];
        handler(entry);
    }
}

- (void)getTimelineEntriesForComplication:(CLKComplication *)complication afterDate:(NSDate *)date limit:(NSUInteger)limit withHandler:(void (^)(NSArray<CLKComplicationTimelineEntry *> * _Nullable))handler
{
    if (complication.family != CLKComplicationFamilyModularLarge ||
        complication.family != CLKComplicationFamilyUtilitarianLarge) {
        handler(nil);
    }
    
    NSMutableArray *timelineEntries = [[NSMutableArray alloc] init];
    NSDate *dateOffset = [date dateByAddingTimeInterval:60 * 60 * 1];
    
    for (NSString *string in self.timelineEntries) {
        if (complication.family == CLKComplicationFamilyUtilitarianLarge) {
            [timelineEntries addObject:[self createUtilitarianLargeTimelineEntryWithHeaderText:@"Top 10" bodyText:string andDate:dateOffset]];
            dateOffset = [dateOffset dateByAddingTimeInterval:60 * 60 * 1];
        }
        
        if (complication.family == CLKComplicationFamilyModularLarge) {
            [timelineEntries addObject:[self createModularLargeTimelineEntryWithHeaderText:@"Top 10" bodyText:string andDate:dateOffset]];
            dateOffset = [dateOffset dateByAddingTimeInterval:60 * 60 * 1];
        }
    }
    
    handler(timelineEntries);
}

#pragma mark - Placeholder Templates 
- (void)getLocalizableSampleTemplateForComplication:(CLKComplication *)complication withHandler:(void (^)(CLKComplicationTemplate * _Nullable))handler
{
    if (complication.family != CLKComplicationFamilyModularLarge ||
        complication.family != CLKComplicationFamilyUtilitarianLarge) {
        handler(nil);
    }
    
    CLKComplicationTemplate *templateToReturn;
    
    if (complication.family == CLKComplicationFamilyModularLarge) {
        CLKComplicationTemplateModularLargeStandardBody *modularLarge = [[CLKComplicationTemplateModularLargeStandardBody alloc] init];
        modularLarge.headerTextProvider = [CLKSimpleTextProvider textProviderWithText:@"Top Song"];
        modularLarge.body1TextProvider = [CLKSimpleTextProvider textProviderWithText:@"-"];
        modularLarge.body2TextProvider = [CLKSimpleTextProvider textProviderWithText:@""];
        templateToReturn = modularLarge;
    }
    
    if (complication.family == CLKComplicationFamilyUtilitarianLarge) {
        CLKComplicationTemplateUtilitarianLargeFlat *utilitarianLarge = [[CLKComplicationTemplateUtilitarianLargeFlat alloc] init];
        utilitarianLarge.textProvider = [CLKSimpleTextProvider textProviderWithText:@"Top Song: ---"];
        templateToReturn = utilitarianLarge;
    }
    
    handler(templateToReturn);
}

#pragma mark - Helper Methods 
- (CLKComplicationTimelineEntry *)createModularLargeTimelineEntryWithHeaderText:(NSString *)header bodyText:(NSString *)body andDate:(NSDate *)date
{
    CLKComplicationTemplateModularLargeStandardBody *modularLarge = [[CLKComplicationTemplateModularLargeStandardBody alloc] init];
    
    modularLarge.headerTextProvider = [CLKSimpleTextProvider textProviderWithText:@"Top Song"];
    
    modularLarge.body1TextProvider = [CLKSimpleTextProvider textProviderWithText:body];
    modularLarge.body2TextProvider = [CLKSimpleTextProvider textProviderWithText:@""];
    
    return [CLKComplicationTimelineEntry entryWithDate:date complicationTemplate:modularLarge];
}

- (CLKComplicationTimelineEntry *)createUtilitarianLargeTimelineEntryWithHeaderText:(NSString *)header bodyText:(NSString *)body andDate:(NSDate *)date
{
    CLKComplicationTemplateUtilitarianLargeFlat *utilitarianLarge = [[CLKComplicationTemplateUtilitarianLargeFlat alloc] init];
    utilitarianLarge.textProvider = [CLKSimpleTextProvider textProviderWithText:body];
    
    return [CLKComplicationTimelineEntry entryWithDate:date complicationTemplate:utilitarianLarge];
}

@end

File: Swift Sample Code/Chapter 7/Top10/Top 10 WatchKit App Extension/ComplicationController.swift
//
//  ComplicationController.swift
//  Top10
//
//  Created by Cory Bohon on 2/14/17.
//  Copyright © 2017 Cory Bohon. All rights reserved.
//

import UIKit
import ClockKit

class ComplicationController: NSObject, CLKComplicationDataSource {
    
    let timelineEntries: Array<String> = ["Never Gonna Give You Up by Rick Astly", "Unlock My iPhone by FBI", "The Screen door Song by Kyle Richter"]
    
    func requestedUpdateDidBegin() {
        
    }

    // MARK: - Timeline Configuration 
    func getSupportedTimeTravelDirections(for complication: CLKComplication, withHandler handler: @escaping (CLKComplicationTimeTravelDirections) -> Void) {
        
        handler(.forward)
    }
    
    func getTimelineStartDate(for complication: CLKComplication, withHandler handler: @escaping (Date?) -> Void) {
        handler(Date())
    }
    
    func getTimelineEndDate(for complication: CLKComplication, withHandler handler: @escaping (Date?) -> Void) {
        handler(Date().addingTimeInterval(6 * 60 * 60))
    }
    
    func getPrivacyBehavior(for complication: CLKComplication, withHandler handler: @escaping (CLKComplicationPrivacyBehavior) -> Void) {
        handler(.showOnLockScreen)
    }
    
    // MARK: - Timeline Population 
    func getCurrentTimelineEntry(for complication: CLKComplication, withHandler handler: @escaping (CLKComplicationTimelineEntry?) -> Void) {
        
        if complication.family == .modularLarge {
            let entry: CLKComplicationTimelineEntry = self.createTimelineEntry(headerText: "Top 10", bodyText: self.timelineEntries.first!, date: Date())
            handler(entry)
        }
    }
    
    func getTimelineEntries(for complication: CLKComplication, after date: Date, limit: Int, withHandler handler: @escaping ([CLKComplicationTimelineEntry]?) -> Void) {
        if complication.family != .modularLarge {
            handler(nil)
        }
        
        var timelineEntries: Array<CLKComplicationTimelineEntry> = Array()
        var dateOffset: Date = date.addingTimeInterval(60 * 60 * 1)
        
        for string: String in self.timelineEntries {
            timelineEntries.append(self.createTimelineEntry(headerText: "Top 10", bodyText: string, date: dateOffset))
            
            dateOffset = dateOffset.addingTimeInterval(60 * 60 * 1)
        }
        
        handler(timelineEntries)
    }
    
    // MARK: - Placeholder Template 
    func getLocalizableSampleTemplate(for complication: CLKComplication, withHandler handler: @escaping (CLKComplicationTemplate?) -> Void) {
        
        if complication.family != .modularLarge {
            handler(nil)
        }
        
        let modularLarge: CLKComplicationTemplateModularLargeStandardBody = CLKComplicationTemplateModularLargeStandardBody()
        modularLarge.headerTextProvider = CLKSimpleTextProvider(text: "Top Song")
        modularLarge.body1TextProvider = CLKSimpleTextProvider(text: "---")
        modularLarge.body2TextProvider = CLKSimpleTextProvider(text: "---")
        
        handler(modularLarge)
    }
    
    // MARK: - Helper Methods 
    func createTimelineEntry(headerText: String, bodyText: String, date: Date) -> CLKComplicationTimelineEntry {
        
        let modularLarge: CLKComplicationTemplateModularLargeStandardBody = CLKComplicationTemplateModularLargeStandardBody()
        modularLarge.headerTextProvider = CLKSimpleTextProvider(text: "Top Song")
        modularLarge.body1TextProvider = CLKSimpleTextProvider(text: bodyText)
        modularLarge.body2TextProvider = CLKSimpleTextProvider(text: "")
        
        return CLKComplicationTimelineEntry(date: date, complicationTemplate: modularLarge)
    }
}
